sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/regexpExpressions.scala

```diff
@@ -41,10 +41,12 @@ trait StringRegexExpression extends Expression
 
   override def dataType: DataType = BooleanType
 
-  // try cache the pattern for Literal
+  // try cache foldable pattern
   private lazy val cache: Pattern = pattern match {
-    case Literal(value: String, StringType) => compile(value)
-    case _ => null
+    case p: Expression if p.foldable =>
```

Comment:
This will consider `'a' + 'b'` from now. I prefer to consider SPARK-30759 as a performance improvement and merge to `master` only.

Reply:
The changes don't impact on behavior in any case, so, they can be considered only as an optimization.